lsp-completion.el

```diff
@@ -397,11 +398,11 @@ The MARKERS and PREFIX value will be attached to each candidate."
            result done?
            (candidates
             (lambda ()
+              ;; Force manual completion action so post completion logic can be run
+              (setq-local company--manual-action t)
```

Comment:
Why is this needed? Post-completion hooks should work without `company--manual-action` (as they did previously), or am I missing something? Removing this line still fixes my `iph` error.

Reply:
In case of there's only one completion result, `company-mode` will do some check of completion result and prefix, if they're the same, the completion popup will not be shown, and `company-cancel` will be triggered with `unique` as result.
In that case, the post-completion will not be run and user will need to manually trigger `company-complete` for the completion to show up, which is kind of annoying.
If we force set `company--manual-action`, the completion popup will always be shown, make the post-completion logic will always be run.

Reply:
The `true.if` in this bug is actually falling on that case too

Reply:
I see. Can't we just run the post-completion hook in `company-after-completion-fn` though, if `RESULT` = `unique`, so that we don't end up relying on implementation details?

Reply:
Aside from this implementation detail, I approve of this PR.

Reply:
From looking at the company source, it seems that if there is a sole match, some company functions should not just complete it, which would be the case for all other company backends, too? Are you sure that `lsp-mode` is the right place to fix that behaviour?

Reply:
If we have to do that to make things work then most likely there is a bug/limitation in company-mode that has to be fixed. Seems like we should include company's maintainer in the loop.

Reply:
Let's loop @dgutov here.
I think that's intentional behavior from `company-mode`. When there's sole match, the completion popup is showing up may seem a little bit silly and intrusive. Thus for automatically invoked completion, not showing the popup (let's call it candidate window) is actually makes sense.
However, in case of manual invoked completion (`company--manual-action` is `t`), the candidate window will always be shown as it's user intention.

For `lsp-mode`, since we actually relied on post-completion, always show the candidate window is better. Thus this hack.

Reply:
IMHO showing/not showing the completion popup should not be related to invoking post-completion. It should happen no matter where the insertion comes from(as long as the candidate is inserted).

Reply:
I think for `company-mode` it's actually related.
Here is the commit (and there's linked discussion) about that behavior https://github.com/company-mode/company-mode/commit/201577cd7ca815b8be2dc62e1f433ee12c1d654f.
`company-mode` auto-completion cannot actually decided if user want to insert the pop-up or just want to type `RET` with current popup. Note that for non `company-tng-mode`, the popup is always preselected.
For `lsp-mode`, user actually relies on completion more so it makes more sense to always showing the popup and let user runs the post-completion

Reply:
This logic is basically for the case when the popup has been automatically shown, but then the user, say, types out the completion in full manually, and then continues typing. It would be weird to do template expansion, etc, in that case.

Reply:
@yyoncho @nbfalcon With @dgutov 's answer, WDYT?
I feel that we should always show the completion popup and let user to run the `exit-fn` properly since `lsp-mode` relies on that to work properly.
Note that when user keeps typing, unless user presses `RET` or `SPC`, the sole match doesn't do anything weird.
Also, with the `company-tng-mode`, user can press `RET` and `SPC` normally since there's no selection by default.